### PR TITLE
boards: arm: Add gd32e103v_eval board

### PR DESCRIPTION
**Pull request by feilongfl**
This PR adds support for GigaDevice [GD32E103VBT6](https://www.gigadevice.com/zh-hans/microcontroller/gd32e103vbt6/) SoC and [GD32E103V-EVAL](https://www.gigadevice.com/products/microcontrollers/gd32-development-tools/gd32-evaluation-boards/) board that is use the SoC.

Currently, the west module dependency points to https://github.com/feilongfl/hal_gd32  (most of these code was download from gigadevice's website, -> Documents & Software -> Firmware Library)
I'm not sure where can I put these code,I found another issues #34971 have gigadevice RISKV device,but I think it's not good to put code there.

At this time, Only `samples/basic/blinky` is work.And lots of code is copy from `stm32`, So some code is not necessary.
I'll remove them later.

### CODE REVIEW

**Comment by carlescufi** [comment]
@feilongfl thank you for this PR. I see that it doesn't require the use of NMSIS anymore, which I think will reassure some of the RISC-V maintainers. Regarding the `.dtsi` files in the https://github.com/feilongfl/hal_gd32 repo, are those present in the upstream repository that this was taken from? If they are not, then you should put them in the main zephyr tree instead.

In general the policy is: if it's written exclusively for Zephyr it should be in the main zephyr tree. If it's part of a cross-RTOS HAL or library that is published separately then it belongs in the external module repo.

**Comment by carlescufi** [comment]
>Currently, the west module dependency points to https://github.com/feilongfl/hal_gd32.

Could you please describe in the main body of this PR where this code comes from?

**Comment by feilongfl** [comment]
> @feilongfl thank you for this PR. Regarding the `.dtsi` files in the https://github.com/feilongfl/hal_gd32 repo, are those present in the upstream repository that this was taken from? If they are not, then you should put them in the main zephyr tree instead.
> 
> In general the policy is: if it's written exclusively for Zephyr it should be in the main zephyr tree. If it's part of a cross-RTOS HAL or library that is published separately then it belongs in the external module repo.

I got it, i will move these `dtsi` code to zephyr prj.

> Could you please describe in the main body of this PR where this code comes from?

most of these code was download from `gigadevice`'s [website](https://www.gigadevice.com/zh-hans/microcontroller/gd32e103vbt6/), -> Documents & Software -> Firmware Library

**Comment by nandojve** [comment]
Hi @carlescufi ,

Do you think Zephyr project can add at least hal_gigadevice ? I have an gd32f403 minimal working too. It was just a test but I can clean up e fix docs and that can be very useful. GigaDevices ARM firmwares are pretty easy to be used with Zephyr.

* [hal_gigadevice](https://github.com/nandojve/hal_gigadevice) - need check licenses
* [gd32f403](https://github.com/nandojve/zephyr/commit/d2c1c92167b5cff551d85ccddfd173909c94e07f) - need clean up and fix docs.

GigaDevices provides a firmware library and tools for all ARM and RISC-V at [GigaDevices Download Site](http://www.gd32mcu.com/en/download/7?kw=GD32F4) .

**Comment by feilongfl** [comment]
I've move `*.dtsi` from module to zephyr repo and migrate to latest version.

**Comment by feilongfl** [comment]
I've migrated this pr to latest main branch. If somewhere need modify, just let me know.

@carlescufi

**Comment by carlescufi** [comment]
@feilongfl or @nandojve could either of you create an RFC issue like [this one](https://github.com/zephyrproject-rtos/zephyr/issues/34971) for the creation of a hal_gd32 or hal_gigadevice repo?

**Comment by nandojve** [comment]
> @feilongfl or @nandojve could either of you create an RFC issue like [this one](https://github.com/zephyrproject-rtos/zephyr/issues/34971) for the creation of a hal_gd32 repo?
> Or should we have a single `hal_gigadevice` repo with both? I think it's probably better.

Hi @carlescufi , it should be a single repo with all gigadevices SoC inside it. I changed name from hal_gd32 to hal_gigadevice to follow name convention because gd32 is the platform name instead vendor name. I can update #38658 scope to reflect RFC description as requested. The #38658 is a proposal to create the initial hal_gigadevice with all 3 SoC firmware base.

**Comment by carlescufi** [comment]
@feilongfl and @nandojve is this PR still applicable now that we finally have a hal_gigadevice?

**Comment by feilongfl** [comment]
> @feilongfl and @nandojve is this PR still applicable now that we finally have a hal_gigadevice?

I am converting `hal_gd32` to `hal_gigadevice` and will split this PR.

**Comment by nandojve** [comment]
Hi @feilongfl ,

Could you rebase and keep only files necessary to introduce board?

**Comment by feilongfl** [comment]
Hi @nandojve 

I've migrated this pr to latest main branch and drop all unnessary file.

need module: https://github.com/zephyrproject-rtos/hal_gigadevice/pull/12

**Comment by nandojve** [comment]
> Hi @nandojve
> 
> I've migrated this pr to latest main branch and drop all unnessary file.
> 
> need module: [zephyrproject-rtos/hal_gigadevice#12](https://github.com/zephyrproject-rtos/hal_gigadevice/pull/12)

Hi @feilongfl ,

You need add a "manifest: update hal_gigadevice" commit, similar first item of #40283. Otherwise there is no access to dependencies from https://github.com/zephyrproject-rtos/hal_gigadevice/pull/12. This is the mandatory trick to get any hal source updated.

I recommend you reorganize in below 4 commits:

- update manifest (update hal_gigadevice)
- dts (new devicetree entries for newer soc)
- soc (new soc entries)
- board (new board series to test newer code)

**Review by gmarull — CHANGES_REQUESTED** [review]
Please, squash all board commits into a single one (after SoC additions). Looks like this PR is missing west.yml update.

File: soc/arm/gigadevice/gd32e10x/Kconfig.defconfig.series
Comment (gmarull) [inline soc/arm/gigadevice/gd32e10x/Kconfig.defconfig.series:10]: this is inconsistent with `SOC_SERIES_GD32E103`, you likely need `SOC_SERIES_GD32E10X`

File: boards/arm/gd32e103v_eval/Kconfig
Comment (gmarull) [inline boards/arm/gd32e103v_eval/Kconfig:8]: this file is not required

File: boards/arm/gd32e103v_eval/doc/index.rst
Comment (gmarull) [inline boards/arm/gd32e103v_eval/doc/index.rst:11]: `.()`? C103V or E103V?

File: boards/arm/gd32e103v_eval/doc/index.rst
Comment (gmarull) [inline boards/arm/gd32e103v_eval/doc/index.rst:42]: This function is not available in this board due to hardware issues, please check ``GD32C103``.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: boards/arm/gd32e103v_eval/Kconfig
Comment (feilongfl) [reply boards/arm/gd32e103v_eval/Kconfig:8]: To be honest, I don't know what this config item means, but I saw it in both `gd32f450i_eval` and `gd32f403z_eval`.
After remove this file, it really compile and run success. So I'll remove this file.

Comment (gmarull) [reply boards/arm/gd32e103v_eval/Kconfig]: it was needed before pinctrl, not anymore ;-)

File: soc/arm/gigadevice/gd32e10x/Kconfig.defconfig.series
Comment (feilongfl) [reply soc/arm/gigadevice/gd32e10x/Kconfig.defconfig.series:10]: Change all `SOC_SERIES_GD32E103` to `SOC_SERIES_GD32E10X`.

File: boards/arm/gd32e103v_eval/doc/index.rst
Comment (feilongfl) [reply boards/arm/gd32e103v_eval/doc/index.rst]: remove `()` and change `C103V` to `E103`

File: boards/arm/gd32e103v_eval/doc/index.rst
Comment (gmarull) [inline boards/arm/gd32e103v_eval/doc/index.rst:42]:
```
``GD32C103``
```

Sphinx uses single quotes for references.

Comment (feilongfl) [reply boards/arm/gd32e103v_eval/doc/index.rst]: I mixed up the formatting of markdown and reStructure.
I have fixed it.

File: boards/arm/gd32e103v_eval/support/openocd.cfg
Comment (nandojve) [inline boards/arm/gd32e103v_eval/support/openocd.cfg:1]: Add `# Copyright (c) 2021, ATL-Electronics` at first line

**Review by nandojve — CHANGES_REQUESTED** [review]
Hi @feilongfl ,

Few comments

File: boards/arm/gd32e103v_eval/doc/index.rst
Comment (gmarull) [reply boards/arm/gd32e103v_eval/doc/index.rst:66]: suggestion: use a structure like this, so that table points to relevant compatibles/Kconfig options

https://github.com/zephyrproject-rtos/zephyr/blob/58234d490c32811860ffc818d662730448c713a7/boards/arm/gd32f450i_eval/doc/index.rst?plain=1#L56-L70

**Review by nandojve — CHANGES_REQUESTED** [review]
```diff
diff --git a/drivers/pinctrl/pinctrl_gd32_afio.c b/drivers/pinctrl/pinctrl_gd32_afio.c
index 409d0c4f01..4f7b4e7852 100644
--- a/drivers/pinctrl/pinctrl_gd32_afio.c
+++ b/drivers/pinctrl/pinctrl_gd32_afio.c
@@ -5,6 +5,7 @@
  */
 
 #include <drivers/pinctrl.h>
+#include <soc.h>
 
 /** AFIO DT node */
 #define AFIO_NODE DT_NODELABEL(afio)
diff --git a/drivers/serial/usart_gd32.c b/drivers/serial/usart_gd32.c
index 8da839684d..ca5910923d 100644
--- a/drivers/serial/usart_gd32.c
+++ b/drivers/serial/usart_gd32.c
@@ -8,6 +8,7 @@
 #include <errno.h>
 #include <drivers/pinctrl.h>
 #include <drivers/uart.h>
+#include <soc.h>
 
 struct gd32_usart_config {
        uint32_t reg;
diff --git a/soc/arm/gigadevice/gd32e10x/soc.c b/soc/arm/gigadevice/gd32e10x/soc.c
index c98d75e83f..80cf3bada5 100644
--- a/soc/arm/gigadevice/gd32e10x/soc.c
+++ b/soc/arm/gigadevice/gd32e10x/soc.c
@@ -5,6 +5,7 @@
 
 #include <device.h>
 #include <init.h>
+#include <soc.h>
 
 static int gd32e10x_soc_init(const struct device *dev)
 {
```

**Comment by feilongfl** [comment]
I've picked wrong commit, fixing.

File: dts/arm/gigadevice/gd32e10x/gd32e10x.dtsi
Comment (nandojve) [inline dts/arm/gigadevice/gd32e10x/gd32e10x.dtsi:72]:
```diff
-               uart3: usart@40004C00 {
+               uart3: usart@40004c00 {
                        compatible = "gd,gd32-usart";
-                       reg = <0x40004C00 0x400>;
+                       reg = <0x40004c00 0x400>;
```

**Review by nandojve — CHANGES_REQUESTED** [review]
You must remove all compile warnings otherwise it may not pass CI

`Warning (simple_bus_reg): /soc/usart@40004C00: simple-bus unit address format error, expected "40004c00"`

**Comment by feilongfl** [comment]
Sorry I accidentally forgot about this PR until it was marked stale, 
I've updated this PR and tested samples/hello_world.

File: soc/arm/gigadevice/gd32e10x/soc.h
Comment (gmarull) [inline soc/arm/gigadevice/gd32e10x/soc.h:11]: please drop `<devicetree.h>`, this is not needed.

File: boards/arm/gd32e103v_eval/gd32e103v_eval.dts
Comment (gmarull) [inline boards/arm/gd32e103v_eval/gd32e103v_eval.dts:28]: note: there are now GPIO and I2C drivers available, can you enable them too? see other boards

Comment (feilongfl) [reply boards/arm/gd32e103v_eval/gd32e103v_eval.dts:28]: OK,I'll check it at this weekend.

File: west.yml
Comment (nandojve) [inline west.yml:76]: Update with new hash and rebase.

**Review by nandojve — CHANGES_REQUESTED** [review]

File: boards/arm/gd32e103v_eval/gd32e103v_eval.dts
Comment (cameled) [reply boards/arm/gd32e103v_eval/gd32e103v_eval.dts:28]: For the I2C driver, you can try the sample `samples/drivers/eeprom`.

**Comment by feilongfl** [comment]
I've enable exti,gpio,timer, dac and i2c.

tested with:
- samples/drivers/eeprom
- samples/drivers/dac
- samples/basic/fade_led
- samples/basic/button

**Review by nandojve — CHANGES_REQUESTED** [review]
Hi @feilongfl !
Two comments only.

File: boards/arm/gd32e103v_eval/gd32e103v_eval_defconfig
Comment (feilongfl) [reply boards/arm/gd32e103v_eval/gd32e103v_eval_defconfig:15]: This change has been removed.

File: samples/drivers/eeprom/boards/gd32e103v_eval.conf
Comment (feilongfl) [reply samples/drivers/eeprom/boards/gd32e103v_eval.conf:3]: I think these conf is already in board overlay.

File: modules/hal_gigadevice/Kconfig
Comment (gmarull) [inline modules/hal_gigadevice/Kconfig:29]: nit: space should not have been removed here

**Review by gmarull — COMMENTED** [review]
LGTM, a couple of minor comments before approval

File: modules/hal_gigadevice/Kconfig
Comment (feilongfl) [reply modules/hal_gigadevice/Kconfig]: fix here

File: boards/arm/gd32e103v_eval/gd32e103v_eval-pinctrl.dtsi
Comment (feilongfl) [reply boards/arm/gd32e103v_eval/gd32e103v_eval-pinctrl.dtsi]: Although the eval board SCL have pull-up resistors, but I'm not sure as IIC master, if SCL needs to be open-drain.
:confused:

Comment (gmarull) [reply boards/arm/gd32e103v_eval/gd32e103v_eval-pinctrl.dtsi]: You still need it, e.g. for https://www.i2c-bus.org/clock-stretching/

**Review by gmarull — APPROVED** [review]

**Review by nandojve — APPROVED** [review]
Hi @feilongfl,
Thank you so much.

**Review by tianrking — COMMENTED** [review]
LGTM

**Pull request merged**
